android/pytorch_android/src/main/java/org/pytorch/Tensor.java

```diff
@@ -0,0 +1,286 @@
+    checkArgument((data.order() == ByteOrder.nativeOrder()),
+        ERROR_MSG_DATA_BUFFER_MUST_HAVE_NATIVE_BYTE_ORDER);
+    return new Tensor_float32(data, dims);
+  }
+
+  public static Tensor newIntTensor(long[] dims, IntBuffer data) {
+    checkArgument(data != null, ERROR_MSG_DATA_BUFFER_NOT_NULL);
+    checkArgument(dims != null, ERROR_MSG_DIMS_NOT_NULL);
+    checkDims(dims);
+    checkDimsAndDataCapacityConsistency(data.capacity(), dims);
+    checkArgument(data.isDirect(), ERROR_MSG_DATA_BUFFER_MUST_BE_DIRECT);
+    checkArgument((data.order() == ByteOrder.nativeOrder()),
+        ERROR_MSG_DATA_BUFFER_MUST_HAVE_NATIVE_BYTE_ORDER);
+    return new Tensor_int32(data, dims);
+  }
+
+  public static Tensor newByteTensor(long[] dims, ByteBuffer data) {
+    checkArgument(data != null, ERROR_MSG_DATA_BUFFER_NOT_NULL);
+    checkArgument(dims != null, ERROR_MSG_DIMS_NOT_NULL);
+    checkDims(dims);
+    checkDimsAndDataCapacityConsistency(data.capacity(), dims);
+    checkArgument(data.isDirect(), ERROR_MSG_DATA_BUFFER_MUST_BE_DIRECT);
+    checkArgument((data.order() == ByteOrder.nativeOrder()),
+        ERROR_MSG_DATA_BUFFER_MUST_HAVE_NATIVE_BYTE_ORDER);
+    return new Tensor_byte(data, dims);
+  }
+
+  private Tensor(long[] dims) {
+    checkDims(dims);
+    this.dims = Arrays.copyOf(dims, dims.length);
+  }
+
+  public static long numElements(long[] dims) {
+    checkDims(dims);
+    int result = 1;
+    for (long dim : dims) {
+      result *= dim;
+    }
+    return result;
+  }
+
+  public float[] getDataAsFloatArray() {
+    throw new IllegalStateException("Tensor of type " + getClass().getSimpleName() + " cannot " +
+        "return data as float.");
+  }
+
+  abstract int getTypeCode();
+
+  Buffer getRawDataBuffer() {
+    throw new IllegalStateException("Tensor of type " + getClass().getSimpleName() + " cannot " +
+        "return raw data buffer.");
+  }
+
+  private static String invalidIndexErrorMessage(int[] index, long dims[]) {
+    return String.format(Locale.US, "Invalid index %s for tensor dimensions %s",
+        Arrays.toString(index), Arrays.toString(dims));
+  }
+
+  static class Tensor_float32 extends Tensor {
+    private final FloatBuffer data;
+
+    Tensor_float32(FloatBuffer data, long[] dims) {
+      super(dims);
+      this.data = data;
+    }
+
+    @Override
+    public float[] getDataAsFloatArray() {
+      data.rewind();
+      float[] arr = new float[data.remaining()];
+      data.get(arr);
+      return arr;
+    }
+
+    @Override
+    int getTypeCode() {
+      return TYPE_CODE_FLOAT32;
+    }
+
+    @Override
+    Buffer getRawDataBuffer() {
+      return data;
+    }
+
+    @Override
+    public String toString() {
+      return String.format("Tensor_float32{dims:%s data:%s}", Arrays.toString(dims),
+          Arrays.toString(getDataAsFloatArray()));
+    }
+  }
+
+  static class Tensor_int32 extends Tensor {
+    private final IntBuffer data;
```

Comment:
btw, is there any difference in storing typed buffer vs just storing the raw buffer all the time? It feel you'd have less code duplication if you just always stored the raw buffer (ByteBuffer I suppose) in the base class and casted on the fly (asIntBuffer seems to create a view on the same memory)